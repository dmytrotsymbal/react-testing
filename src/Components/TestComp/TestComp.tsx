import { useEffect, useState } from 'react'

type Props = {}
const TestComp = (props: Props) => {
    const [data, setData] = useState<null | any>(null)

    const [toggle, setToggle] = useState<boolean>(false)

    const handleToggle = () => {
        setToggle(!toggle)
    }

    useEffect(() => {
        setTimeout(() => {
            setData({})
        }, 100)
    }, [])

    const [value, setValue] = useState<string>('')

    return (
        <div className="TestComp" data-testid="testComp">
            {data && <h6 style={{ color: 'red' }}>randomText</h6>}
            <h1>TestComp</h1>
            <button
                onClick={handleToggle}
                data-testid="toggleBtn"
                className="btn"
            >
                Click me!
            </button>
            {toggle === true && <h6 data-testid="toggleElement">toggle</h6>}

            <hr />
            <h3 data-testid="valueH3Element">{value}</h3>
            <input
                type="text"
                placeholder="Type here"
                onChange={(e) => setValue(e.target.value)}
            />
        </div>
    )
}
export default TestComp
