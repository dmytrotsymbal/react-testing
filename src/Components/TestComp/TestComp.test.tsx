import { screen, render, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom/extend-expect'
import TestComp from './TestComp'

test('component contains elements', () => {
    render(<TestComp />)
    const titleElement = screen.getByText(/TestComp/i)
    expect(titleElement).toBeInTheDocument()

    const buttonElement = screen.getByRole('button')
    expect(buttonElement).toBeInTheDocument()
    expect(buttonElement).toHaveClass('btn')

    const inputElement = screen.getByPlaceholderText(/Type here/i)
    expect(inputElement).toBeInTheDocument()
    expect(inputElement).toMatchSnapshot()
})

test('randomText render', async () => {
    render(<TestComp />)
    const randomTextElement = await screen.findByText(/randomText/i)
    expect(randomTextElement).toBeInTheDocument()
    expect(randomTextElement).toHaveStyle({
        color: 'red',
    })
})

test('CLICK EVENT', async () => {
    render(<TestComp />)
    const toggleButton = screen.getByTestId('toggleBtn')
    expect(toggleButton).toBeInTheDocument()

    //-----------------------------------------------------------------------
    //OnClick event
    expect(screen.queryByTestId('toggleElement')).toBeNull()
    fireEvent.click(toggleButton)
    expect(screen.getByTestId('toggleElement')).toBeInTheDocument()
    fireEvent.click(toggleButton)
    expect(screen.queryByTestId('toggleElement')).toBeNull()
})

test('INPUT EVENT', async () => {
    render(<TestComp />)
    const inputElement = screen.getByPlaceholderText(/Type here/i)
    expect(inputElement).toBeInTheDocument()

    //-----------------------------------------------------------------------
    //INPUT event
    expect(screen.getByTestId('valueH3Element')).toContainHTML('')

    // fireEvent.input(inputElement, { target: { value: 'test123' } })
    userEvent.type(inputElement, 'test123')
    expect(screen.getByTestId('valueH3Element')).toContainHTML('test123')
})

test('component snapshot', () => {
    const { asFragment } = render(<TestComp />)
    expect(asFragment()).toMatchSnapshot()
})
