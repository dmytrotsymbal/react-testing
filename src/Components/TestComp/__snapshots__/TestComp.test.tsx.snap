// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component contains elements 1`] = `
<input
  placeholder="Type here"
  type="text"
/>
`;

exports[`component snapshot 1`] = `
<DocumentFragment>
  <div
    class="TestComp"
    data-testid="testComp"
  >
    <h1>
      TestComp
    </h1>
    <button
      class="btn"
      data-testid="toggleBtn"
    >
      Click me!
    </button>
    <hr />
    <h3
      data-testid="valueH3Element"
    />
    <input
      placeholder="Type here"
      type="text"
    />
  </div>
</DocumentFragment>
`;
